pkg/oc/cli/cmd/rollout/latest.go

```diff
@@ -126,18 +142,73 @@ func (o *RolloutLatestOptions) Complete(f *clientcmd.Factory, cmd *cobra.Command
 
 func (o RolloutLatestOptions) Validate() error {
 	if len(o.infos) != 1 {
-		return errors.New("a deployment config name is required.")
+		return errors.New("a deployment config name is required")
 	}
 	return nil
 }
 
-func (o RolloutLatestOptions) RunRolloutLatest() error {
-	info := o.infos[0]
-	config, ok := info.Object.(*deployapi.DeploymentConfig)
+// rolloutDeployments rollouts the latest version for deployments.
+func (o RolloutLatestOptions) rolloutDeployment(d *kextensions.Deployment) error {
+	out, ok := d.GetAnnotations()[triggerapi.TriggerAnnotationKey]
 	if !ok {
-		return fmt.Errorf("%s is not a deployment config", info.Name)
+		return fmt.Errorf("deployment %s does not have any triggers defined", d.Name)
+	}
+	triggers := []triggerapi.ObjectFieldTrigger{}
+	if err := json.Unmarshal([]byte(out), &triggers); err != nil {
+		return err
+	}
+	var (
+		errors                []error
+		hasPausedImageTrigger bool
+		hasUpdatedImages      bool
+	)
+	obj, err := kapi.Scheme.DeepCopy(d)
+	if err != nil {
+		return err
 	}
+	dCopy := obj.(*kextensions.Deployment)
+	for _, trigger := range triggers {
+		if !trigger.Paused {
+			continue
+		}
+		container, remainder, err := annotations.ContainerForObjectFieldPath(dCopy, trigger.FieldPath)
+		if err != nil || remainder != "image" {
+			continue
+		}
+		hasPausedImageTrigger = true
+		spec, err := triggerutil.ResolveLatestTriggerImagePullSpec(o.imageClient, dCopy.Namespace, trigger.From)
+		if err != nil {
+			errors = append(errors, fmt.Errorf("unable to resolve image %s for container %q: %v", trigger.From.Name, container.GetName(), err))
+			continue
+		}
+		if container.GetImage() != spec.String() {
+			container.SetImage(spec.String())
+			hasUpdatedImages = true
+		}
+	}
+	if !hasPausedImageTrigger {
+		return fmt.Errorf("deployment %q has no paused triggers", dCopy.Name)
+	}
+	if !hasUpdatedImages {
+		return fmt.Errorf("deployment %q already runs the latest images", dCopy.Name)
+	}
+	if len(errors) > 0 {
+		return fmt.Errorf("deployment %q: %v", dCopy.Name, utilerrors.NewAggregate(errors))
+	}
+	updatedDeployment, err := o.kc.Extensions().Deployments(dCopy.Namespace).Update(dCopy)
```

Comment:
I would suggest using PATCH for 2 reasons:
1. you don't have to retry in case of collision
2. I think that by doing client side update due to yaml -> object -> yaml conversions you will strip any fields unknown to you that might be present in newer version OpenShift

Reply:
@tnozicka i'm really against patch here, if this fails because something change the paused this will race with the image trigger controller and gives you results you don't want probably. I don't think 2) can happen as the upstream client will deal with conversions on its own.

Reply:
@mfojtik actually I've been there with deployer image - if there is new (optional) field in OpenShift and you use client that's compiled with that object in the old version not having it, you loose it with update. And as that field is optional (+omitempty) server has no way of knowing if you want it unspecified or you don't know about it.

You can avoid your concern by including resourceVersion in PATCH

Reply:
we don't use the PATCH for instantiate (for good reason) and also not for oc set image (which ultimately this is just copying...).